### PR TITLE
Implement state_vector method on QubitStateVector

### PR DESCRIPTION
**Pull request by timmysilv**
Replaces #3635

**Context:**
State-prep methods should implement (aka inherit from) the new `StatePrep` interface, so this PR does that.

**Description of the Change:**
QubitStateVector inherits from StatePrep, and implements the `state_vector` method. It essentially just expands over the provided `wire_order` and reshapes the parameter from flat to `(2,) * num_wires`

**Benefits:**
QubitStateVector will be usable to prepare a state in the new python simulator device (once #3683 is done)

**Possible Drawbacks:**
N/A

Question for reviewers that I'm now realizing: The wire expansion thing might be common, and should maybe follow the existing `matrix/compute_matrix` pattern so the parent StatePrep class can do the expansion, and the child classes will be simpler. Should I make that change? I'm thinking yes.

### CODE REVIEW

File: doc/releases/changelog-dev.md
Comment (albi3ro) [inline doc/releases/changelog-dev.md:325]:
```suggestion
* `QubitStateVector` now implements the `StatePrep` interface.
  [(#3685)](https://github.com/PennyLaneAI/pennylane/pull/3685)
```

**Review by albi3ro — COMMENTED** [review]
Trial case that's not behaving as expected:
```
>>> op = qml.QubitStateVector(np.array([0, 1, 0, 0]), wires=[1,0])
>>> op.state_vector(wire_order=(1,0))
array([[0., 1.],
       [0., 0.]])
>>> op.state_vector(wire_order=(0,1))
array([[0., 1.],
       [0., 0.]])
```

Good to compare behaviour with `DefaultQubit._apply_state_vector` during developement.

**Comment by timmysilv** [comment]
@albi3ro good point! I've updated it to handle wire order now. I kinda gave up comparing with default.qubit because it doesn't handle a different set of wires from the device itself, so this method actually provides _additional_ functionality! (at the cost of an assumption that the extra wires are zero)

File: pennylane/ops/qubit/state_preparation.py
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:178]:
```suggestion
        if wire_order is None or Wires(wire_order) == self.wires:
```

File: pennylane/ops/qubit/state_preparation.py
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:181]: Any reason not to keep the term `wire_order`? I think `wire_order` helps distinguish it a bit from "wires" being the operation `self.wires`.

Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:181]: it's a slight difference but not a problem, i'll update accordingly!

Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:181]: on second thought, I only need it to be a `Wires` instance for this check, so I'll just cast it there

File: pennylane/ops/qubit/state_preparation.py
Comment (rmoyard) [inline pennylane/ops/qubit/state_preparation.py:177]: I think you need the precision part of the functions arguments because user can requester their own c_dtype in the device and it will not always be complex 128.

File: pennylane/ops/qubit/state_preparation.py
Comment (rmoyard) [inline pennylane/ops/qubit/state_preparation.py:175]: General question about the device rework: what do we want to do about batching? Because this function should be compatible with it at the end of the rework. In my opinion it is easier to make sure to tackle this issue in each separate PR instead of coming back to it later. @albi3ro

File: tests/ops/qubit/test_state_prep.py
Comment (rmoyard) [inline tests/ops/qubit/test_state_prep.py:111]: Can we add a example with many reordering necessary?

**Review by rmoyard — COMMENTED** [review]
Hi @timmysilv, it looks good to me :100: A couple of questions before approving it

File: pennylane/ops/qubit/state_preparation.py
Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:177]: I wasn't entirely sure how to handle this, because it's not tied to a device anymore. `math.cast` seemed to be the best way that also works with all interfaces. I didn't want to force any dtype here, nor did I want to force users to manually specify floats or complex values (why shouldn't `[1,0,0,0]` be a valid 2-qubit state?), but the `norm` check below requires it and that's why I added it. @albi3ro perhaps we should expand the `state_vector` function prototype to include `dtype` as a kwarg with `np.complex128` as the default (it dispatches to other interfaces properly)

File: tests/ops/qubit/test_state_prep.py
Comment (timmysilv) [reply tests/ops/qubit/test_state_prep.py:111]: I did this below in a separate test (`test_QubitStateVector_reordering`) because the $\ket{01}$ state used in this test is a little trivial for convincing the reader of my point. lmk if you'd like more/different testing for it

File: pennylane/ops/qubit/state_preparation.py
Comment (albi3ro) [reply pennylane/ops/qubit/state_preparation.py:175]: Exactly. Batching is a "nice to have".  If it would be straightforward to implement right now, go ahead and add it.  But I'd like to try and get as much basic behaviour in as possible right now and then come back to add more feature coverage.

File: pennylane/ops/qubit/state_preparation.py
Comment (albi3ro) [reply pennylane/ops/qubit/state_preparation.py:177]: Yes, we should be able to specify the `dtype` for a simulation.

File: pennylane/ops/qubit/state_preparation.py
Comment (albi3ro) [inline pennylane/ops/qubit/state_preparation.py:185]: Should we do this on initialization instead? "Fail fast".

Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:185]: we didn't before, but I think we should. I didn't try to in this PR because we still use the parent `__init__`, but I think I'll do it in this PR

File: pennylane/ops/qubit/state_preparation.py
Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py:177]: kk. fortunately, adding an `__init__` to this class will mean I don't have to do this casting when computing `state_vector` so it's less of a concern in this PR (you can set dtype by setting the dtype of the parameter you provide), but in a new PR I'll add `dtype` to `state_vector`.

File: pennylane/ops/qubit/state_preparation.py
Comment (timmysilv) [reply pennylane/ops/qubit/state_preparation.py]: took a crack at it in new commits

File: tests/tape/test_tape.py
Comment (rmoyard) [inline tests/tape/test_tape.py:863]: Was it changed because an error was raised or because you wanted to test the reordering it inside a tape?

**Review by rmoyard — APPROVED** [review]
It looks good to me :100: But do not merge before a second approval pls

File: tests/tape/test_tape.py
Comment (timmysilv) [reply tests/tape/test_tape.py:863]: error was raised. if wires has length $n$, the provided state must have length $2^n$

Comment (rmoyard) [reply tests/tape/test_tape.py:863]: Before your could apply it on a subset of wires and assume everything else is zero, right?

Comment (timmysilv) [reply tests/tape/test_tape.py:863]: nope, the logic used to be in`DefaultQubit._apply_state_vector()` (which was called in `DefaultQubit.apply()`), meaning the error would only be raised when it is being applied by a device. Now, I raise the error at construction-time in `QubitStateVector.__init__()` because it was not a legitimate Operator at that point.

As for the everything-else-is-zero assumption, applying `QubitStateVector` _used to_ fail if it didn't touch all wires on the device it's being applied to. With this PR, it makes the zero-assumption for missing wires as an added functionality. I started implementing that before realizing that it was an error on the old device, but now that I type it out, I wonder if maybe we don't want to add that feature?

**Review by albi3ro — APPROVED** [review]
Looks good to me!

**Pull request merged**
